### PR TITLE
afsmtp: recipients are templatable now.

### LINKED ISSUE #197: Enhancement: afsmtp: make to() templateable
I'd like to make the to() parameter of smtp to be dynamic, by feeding it a macro.
In the same manner as subject() and body()

### PR DESCRIPTION
**Pull request by hkrist**
request: #197
closes#197
fixes #197 
Signed-off-by: Kristof Havasi <kristof.havasi95@gmail.com>

### CODE REVIEW

**Comment by gregory094** [comment]
+1

File: modules/afsmtp/afsmtp.c [hunk 1]
```diff
@@ -45,6 +45,7 @@ typedef struct
 {
   gchar *phrase;
   gchar *address;
+  LogTemplate *addr_tmpl;
   afsmtp_rcpt_type_t type;
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Do we need both the address and addr_tmpl here?

**Comment by bazsi** [comment]
Pls use a better commit message. First line is a summary, subsequent lines should contain rationale, further information. Pls look at the git history for examples.

File: modules/afsmtp/afsmtp.c [hunk 2]
```diff
@@ -134,6 +135,7 @@ afsmtp_dd_set_from(LogDriver *d, const gchar *phrase, const gchar *mbox)
 
   g_free(self->mail_from->phrase);
   g_free(self->mail_from->address);
+
   self->mail_from->phrase = afsmtp_wash_string(g_strdup(phrase));
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Please don't include such unrelated whitespace changes.

File: modules/afsmtp/afsmtp.c [hunk 3]
```diff
@@ -225,12 +227,20 @@ afsmtp_dd_format_persist_name(LogThrDestDriver *d)
...
  */
+
 static void
-afsmtp_dd_msg_add_recipient(AFSMTPRecipient *rcpt, smtp_message_t message)
+afsmtp_dd_msg_add_recipient(AFSMTPRecipient *rcpt, gpointer user_data)
 {
-  gchar *hdr;
+  AFSMTPDriver *self = ((gpointer *)user_data)[0];
+  LogMessage *msg = ((gpointer *)user_data)[1];
+  smtp_message_t message = ((gpointer *)user_data)[2];
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Perhaps split the user data parsing and the actual code to perform the operation. The body should be extracted to a separate function with typesafe arts.

Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Args

File: modules/afsmtp/afsmtp.c [hunk 4]
```diff
@@ -382,24 +393,28 @@ __build_message(AFSMTPDriver *self, LogMessage *msg, smtp_session_t session)
 
   message = smtp_add_message(session);
 
-  smtp_set_reverse_path(message, self->mail_from->address);
+  log_template_format(self->mail_from->addr_tmpl, msg, &self->template_options, LTZ_SEND,
+                      self->super.seq_num, NULL, self->str);
+  smtp_set_reverse_path(message, afsmtp_wash_string(self->str->str));
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Hmm probably reverse_path should take care about  template formatting, and not open coding it numerous times.

How does memory ordership go? What frees the return value of wash_string?

File: modules/afsmtp/afsmtp.c [hunk 5]
```diff
@@ -536,6 +551,31 @@ afsmtp_worker_thread_deinit(LogThrDestDriver *d)
...
+	  hdr = "Cc";
+	  break;
+	case AFSMTP_RCPT_TYPE_REPLY_TO:
+	  hdr = "Reply-To";
+	  break;
+	default:
+	  return;
+	}
+      rcpt->addr_tmpl = log_template_new(cfg, hdr);
+      log_template_compile(rcpt->addr_tmpl, rcpt->address, NULL);
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Error handling of template compilation?

File: modules/afsmtp/afsmtp.c [hunk 6]
```diff
@@ -636,6 +678,11 @@ afsmtp_dd_init(LogPipe *s)
       self->body_tmpl = log_template_new(cfg, "body");
       log_template_compile(self->body_tmpl, self->body, NULL);
     }
+  if (!self->mail_from->addr_tmpl)
+    {
+      self->mail_from->addr_tmpl = log_template_new(cfg, "from");
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Associating a name to this seems to be wrong for me

File: modules/afsmtp/afsmtp.c [hunk 7]
```diff
@@ -626,6 +666,8 @@ afsmtp_dd_init(LogPipe *s)
     return FALSE;
 
   g_list_foreach(self->headers, (GFunc)afsmtp_dd_init_header, cfg);
+  g_list_foreach(self->rcpt_tos, (GFunc)afsmtp_dd_init_recipient, cfg);
+
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: This seems to be an independent change to me, and as such should go to a separate patch

File: modules/afsmtp/afsmtp.c (hunk 5 above)
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Template name, that field is normally used by named templates in the config and not otherwise. Can be null.

**Comment by bazsi** [comment]
-1, this needs some polish before going in.

File: modules/afsmtp/afsmtp.c (hunk 1 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: Well, all of Subject(), Body() and Header() are following this pattern. All of them contains its own string (from the parsed syslogconf) and its LogTemplate struct for the compiled string.
Of course the parsing functions could compile the strings from the syslogconf directly into the LogTemplate struct and that would eventuate that the recipient structs wouldn't need the gchar *address, but as I mentioned that would break the consistency between the smtp templating methods.

File: modules/afsmtp/afsmtp.c (hunk 2 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: Ok, thanks i'm reworking that.

File: modules/afsmtp/afsmtp.c (hunk 7 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: This new function added "afsmtp_dd_init_recipient" is an essential step in the template compiling, 
it works like the afsmtp_dd_init_header method(one line above it) , it compiles(changes the $macro parts) the recipient's address strings into their addr_tmpl.

File: modules/afsmtp/afsmtp.c (hunk 4 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: The next log_template_format function frees the string and it is called after line396  multiple times

Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: Could  you be more precise what you meant on your first sentence? From() shouldn't take macros?

File: modules/afsmtp/afsmtp.c (hunk 2 above)
Comment (gregory094) [inline modules/afsmtp/afsmtp.c]: Take care of functions that return pointer to dynamically allocated memory. (E.g.: g_strdup, wash_string) These gchar* strings should be freed when no longer needed. See glib documentation: [here](https://developer.gnome.org)

File: modules/afsmtp/afsmtp.c (hunk 4 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: log_template_format is called multiple times after this, it takes care of freeing (and adding new value to) the self->str.

File: modules/afsmtp/afsmtp.c (hunk 1 above)
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: If we are not using the string representation  I would get rid off those everywhere.

Also, LogTemplate contains the template string in addition to the compiled representation.

File: modules/afsmtp/afsmtp.c (hunk 4 above)
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: What is self->str? The message being sent?

Perhaps a rename would be due here. Is it just a temp buffer?

File: modules/afsmtp/afsmtp.c (hunk 1 above)
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: One further thought: what happens if template compilation fails in either parameters. We should fail the startup as well.

File: modules/afsmtp/afsmtp.c (hunk 2 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: It's still not "my code" though, the only change was here the whitespace(which i will fix) BUT they are free-d somewhere at line 700 so that isn't a problem either.

Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: free-d around line 700. (This still not my code though, only change was the whitespace, which i'll fix)

File: modules/afsmtp/afsmtp.c (hunk 4 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: yes as i noticed it's for temporary uses, mostly for containing the fully templatecompiled strings

**Comment by bazsi** [comment]
Have all nitpicks been addressed in the recent changes? Do you consider this mergable?
Thanks.

**Comment by hkrist** [comment]
Added some polish to it, appeasing the requirements, which was discussed here. In the last commit I got rid of the string representations everywhere of the templatable strings as it was asked, it required some reorganization of the module initializing functions, but the AFSMTPDriver got more 
space sparing.
The templating of to(), bcc(), cc(), from(), reply-to() works just like it does with header(), subject() and body(). e.g from("sender" "usr@$HOST"), to("recipient" "usr@$SOMEMACRO"), so the second string(address) of the bodies got templatable (if only one string is used e.g. to("usr@$SOMEMACRO"), than it is templatable).
I consider this mergable.

**Comment by gregory094** [comment]
@bazsi what do u think about this? I've tried it out and it's working.  @hkrist thinks it's mergable I'll check the code, too.

File: modules/afsmtp/afsmtp-grammar.ym [hunk 8]
```diff
@@ -75,25 +75,25 @@ afsmtp_options
...
         : KW_HOST '(' string ')'		{ afsmtp_dd_set_host(last_driver, $3); free($3); }
         | KW_PORT '(' LL_NUMBER ')'		{ afsmtp_dd_set_port(last_driver, $3); }
-	| KW_SUBJECT '(' string ')'	 	{ afsmtp_dd_set_subject(last_driver, $3); free($3); }
-	| KW_BODY '(' string ')'		{ afsmtp_dd_set_body(last_driver, $3); free($3); }
+	| KW_SUBJECT '(' string ')'	 	{ afsmtp_dd_set_subject(last_driver, $3, configuration); free($3); }
+	| KW_BODY '(' string ')'		{ afsmtp_dd_set_body(last_driver, $3, configuration); free($3); }
 	| KW_HEADER '(' string string ')'	{
-		afsmtp_dd_add_header(last_driver, $3, $4);
+		afsmtp_dd_add_header(last_driver, $3, $4, configuration);
 		free($3); free($4);
```
Comment (bazsi) [inline modules/afsmtp/afsmtp-grammar.ym]: Please pass the configuration object to the constructor and don't specify it for all of these calls.

File: modules/afsmtp/afsmtp.c [hunk 9]
```diff
@@ -119,51 +116,56 @@ afsmtp_dd_set_port(LogDriver *d, gint port)
...
+afsmtp_dd_set_subject(LogDriver *d, const gchar *subject, GlobalConfig *cfg)
 {
   AFSMTPDriver *self = (AFSMTPDriver *)d;
 
-  g_free(self->subject);
-  self->subject = g_strdup(subject);
+  log_template_unref(self->subject_tmpl);
+  self->subject_tmpl = log_template_new(cfg, "subject");
+  self->subject_tmpl->template = g_strdup(subject);
 }
```
Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: Hmmm why do you need to poke into that member?

This should probably be a log_template_compile call. Or even better use the template_content rule in the grammar instead of a string, that will give you proper error handling as well.

What happens if an invalid template is specifiied?

**Comment by bazsi** [comment]
It could use some polish yet

File: modules/afsmtp/afsmtp.c (hunk 9 above)
Comment (hkrist) [inline modules/afsmtp/afsmtp.c]: I poked into that member, because you suggested that we should get rid of the string representations of the templated strings. 
In my latest commit bc1a25203612c60faefe5fe61265908f51e62568 I used the template_content rule and it takes care of calling log_template_new, log_template_compile and also takes care of error handling, so if the log_template_compile function runs to an invalid template, syslog-ng fails the startup.

Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: -- 
Bazsi

On Mon, Jul 6, 2015 at 12:01 PM, hkrist <notifications@github.com> wrote:

> In modules/afsmtp/afsmtp.c
> <https://github.com/balabit/syslog-ng/pull/419#discussion_r33920366>:
>
> >  {
> >    AFSMTPDriver *self = (AFSMTPDriver *)d;
> >
> > -  g_free(self->subject);
> > -  self->subject = g_strdup(subject);
> > +  log_template_unref(self->subject_tmpl);
> > +  self->subject_tmpl = log_template_new(cfg, "subject");
> > +  self->subject_tmpl->template = g_strdup(subject);
> >  }
>
> I poked into that member, because you suggested that we should get rid of
> the string representations of the templated strings.
>

that string is managed by LogTemplate internally, it is where it saves the
compiled template. if you call log_template_compile() it will save its
input there. Isn't that enough?

Also, you don't need to specify a name to log_template_new(), just pass a
NULL. LogTemplate should not have a name, the only reason it does is that
the same struct is used as both the template itself and a configuration
object (e.g. template "foobar" in the confiuguration) which requires a
name. It should be split, but as long as it isn't we don't need to use name
anywhere, except for the configuration parser which parses the top-level
template statement.


> In my latest commit bc1a252
> <https://github.com/balabit/syslog-ng/commit/bc1a25203612c60faefe5fe61265908f51e62568>
> I used the template_content rule and it takes care of calling
> log_template_new, log_template_compile and also takes care of error
> handling, so if the log_template_compile function runs to an invalid
> template, syslog-ng fails the startup.
>
I'll check that. Thanks,

Comment (bazsi) [inline modules/afsmtp/afsmtp.c]: The patch bc1a252 looks great at first sight. I think it does leak a
LogTemplate reference though. template_content would return a reference
which is not freed anywhere.

PS: I know that the free/destructor part of the parser sucks, it's very
easy to make a mistake there.


-- 
Bazsi

**Comment by gregory094** [comment]
I reviewed this patch on the weekend. It seems good but it need some polishments I'll write some of my notes to the code. 
I moved this to "Waiting for review".

File: modules/afsmtp/afsmtp.c [hunk 10]
```diff
@@ -225,13 +224,11 @@ afsmtp_dd_format_persist_name(LogThrDestDriver *d)
 /*
  * Worker thread
  */
+
 static void
-afsmtp_dd_msg_add_recipient(AFSMTPRecipient *rcpt, smtp_message_t message)
+afsmtp_dd_msg_add_recipient_header(smtp_message_t message, AFSMTPRecipient *rcpt, AFSMTPDriver *self)
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please make `AFSMTPDriver *self` parameter as the first parameter. It's a convention to use the `self` parameter as the first one.

File: modules/afsmtp/afsmtp.c [hunk 11]
```diff
@@ -246,18 +243,38 @@ afsmtp_dd_msg_add_recipient(AFSMTPRecipient *rcpt, smtp_message_t message)
...
       return;
     }
-  smtp_set_header(message, hdr, rcpt->phrase, rcpt->address);
+
+  smtp_set_header(message, hdr, rcpt->phrase, afsmtp_wash_string (self->str->str));
   smtp_set_header_option(message, hdr, Hdr_OVERRIDE, 1);
 }
 
 static void
+afsmtp_dd_msg_add_recipient_with_template(smtp_message_t message, AFSMTPDriver *self, LogTemplate *addr_tmpl, LogMessage *msg)
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please make `AFSMTPDriver *self` parameter as the first parameter. It's a convention to use the `self` parameter as the first one.

File: modules/afsmtp/afsmtp.c [hunk 12]
```diff
@@ -37,14 +37,13 @@
 typedef struct
 {
   gchar *name;
-  gchar *template;
-  LogTemplate *value;
+  LogTemplate *hdr_tmpl;
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please rename `hdr_tmpl`  to `template`. It's a redundant information.

File: modules/afsmtp/afsmtp.c [hunk 13]
```diff
@@ -37,14 +37,13 @@
...
 } AFSMTPHeader;
 
 typedef struct
 {
   gchar *phrase;
-  gchar *address;
+  LogTemplate *addr_tmpl;
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please rename `addr_tmpl`  to `template`. It's a redundant information.

File: modules/afsmtp/afsmtp.c [hunk 14]
```diff
@@ -57,11 +56,9 @@ typedef struct
   gchar *host;
   gint port;
 
-  gchar *subject;
   AFSMTPRecipient *mail_from;
   GList *rcpt_tos;
   GList *headers;
-  gchar *body;
 
   LogTemplate *subject_tmpl;
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please rename `subject_tmpl` to `subject_template`. It would make the code consistent (if you have already renamed the fields in `AFSMTPRecipient` and `AFSMTPHeader` )

File: modules/afsmtp/afsmtp.c [hunk 15]
```diff
@@ -57,11 +56,9 @@ typedef struct
...
   LogTemplate *body_tmpl;
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please rename `subject_tmpl` and `body_tmpl` to `subject_template` and `body_template`. It would make the code consistent (if you have already renamed the fields in `AFSMTPRecipient` and `AFSMTPHeader` )

File: modules/afsmtp/afsmtp.c (hunk 11 above)
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: This function should be renamed to `_smtp_message_add_recipient_from_template()`.

File: modules/afsmtp/afsmtp.c [hunk 16]
```diff
@@ -382,24 +399,28 @@ __build_message(AFSMTPDriver *self, LogMessage *msg, smtp_session_t session)
...
-  smtp_set_reverse_path(message, self->mail_from->address);
+  log_template_format(self->mail_from->addr_tmpl, msg, &self->template_options, LTZ_SEND,
+                      self->super.seq_num, NULL, self->str);
+  smtp_set_reverse_path(message, afsmtp_wash_string(self->str->str));
 
   /* Defaults */
   smtp_set_header(message, "To", NULL, NULL);
   smtp_set_header(message, "From", NULL, NULL);
 
+
```
Comment (ihrwein) [inline modules/afsmtp/afsmtp.c]: Please remove this empty line.

**Comment by ihrwein** [comment]
For the note, Gergo sit next to me during the review :)

File: modules/afsmtp/afsmtp.c (hunk 16 above)
Comment (dnsjts) [inline modules/afsmtp/afsmtp.c]: Removed in d3b4a7b2313e980b21cf6c1e6d55fe08ae722919.

File: modules/afsmtp/afsmtp.c (hunk 11 above)
Comment (dnsjts) [inline modules/afsmtp/afsmtp.c]: Renamed in d3b4a7b2313e980b21cf6c1e6d55fe08ae722919.

File: modules/afsmtp/afsmtp.c (hunk 12 above)
Comment (dnsjts) [inline modules/afsmtp/afsmtp.c]: Renamed in d3b4a7b2313e980b21cf6c1e6d55fe08ae722919.

File: modules/afsmtp/afsmtp.c (hunk 13 above)
Comment (dnsjts) [inline modules/afsmtp/afsmtp.c]: Renamed in d3b4a7b2.

File: modules/afsmtp/afsmtp.c (hunk 15 above)
Comment (dnsjts) [inline modules/afsmtp/afsmtp.c]: Renamed in d3b4a7b.

**Comment by dnsjts** [comment]
:-1: , the mentioned issues have been fixed, but additional issues have been introduced, fix them please.

Additionaly, please try to reduce the size of a commit (of course within reason), because it is easier to review/test a commit, which does only one thing. (And merging two commits is a bit easier, than splitting one.)

**Comment by gregory094** [comment]
@hkrist updated his PR
I think it's fair enough now. Could u look at it @dnsjts?

**Comment by dnsjts** [comment]
:+1:

**Comment by lbudai** [comment]
some whitespace issues around grammar file, but otherwise: :+1:

**Comment by ihrwein** [comment]
I'll open an issue for the whitespace handling in the grammar file, but otherwise ok. Thanks for your contribution! :+1:

**Comment by ihrwein** [comment]
I've opened an issue about whitespaces: https://github.com/balabit/syslog-ng/issues/638
